src/Service/StateHandler.php

```diff
@@ -0,0 +1,229 @@
+<?php
+/**
+ * Copyright 2017 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\GsspBundle\Service;
+
+use Surfnet\GsspBundle\Exception\RuntimeException;
+use Surfnet\SamlBundle\SAML2\AuthnRequest;
+use Surfnet\SamlBundle\SAML2\ReceivedAuthnRequest;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Knows and preserves the integrity of the GSSP application state.
+ */
+final class StateHandler implements StateHandlerInterface
+{
+    const REQUEST_ID = 'request_id';
+    const REQUEST_STEPUP_ID = 'stepup_request_id';
+    const REQUEST_SERVICE_PROVIDER = 'service_provider';
+    const REQUEST_RELAY_STATE = 'relay_state';
+
+    const REQUEST_TYPE = 'request_type';
+    const REQUEST_TYPE_AUTHENTICATION = 'request_type_authentication';
+    const REQUEST_TYPE_REGISTRATION = 'request_type_registration';
+
+    /**
+     * This can be the request identity name id or the response registered named id.
+     */
+    const NAME_ID = 'name_id';
+
+    const RESPONSE_TYPE = 'response_type';
+    const RESPONSE_TYPE_USER_REGISTERED = 'response_type_user_registered';
+    const RESPONSE_TYPE_USER_AUTHENTICATED = 'response_type_user_authenticated';
+    const RESPONSE_TYPE_ERROR = 'response_type_error';
+
+    const RESPONSE_ERROR_SUB_CODE = 'error_response_sub_code';
+    const RESPONSE_ERROR_MESSAGE = 'error_response_message';
+
+    private $store;
+
+    public function __construct(ValueStore $store)
+    {
+        $this->store = $store;
+    }
+
+    public function saveRegistrationRequest(Request $request, ReceivedAuthnRequest $authnRequest, $stepupRequestId)
```

Comment:
Wasn't one of the points of implementing the `saveRegistrationRequest` and `saveAuthenticationRequest` in the State handler to be able to protect the integrity in a more elegant way? 

I expected to see more input validation on the input of these methods. Maybe this is not needed as their state is validated elsewhere. Just being curious here.

Reply:
Good question. 

1. The controller validates input data like the AuthnRequest. 
2. This StateHandler protects the integrity of the current state and when changes can happen to it. Not if if the given variables are correct. I did add some minor assertions to know that the given variables are of the correct scaler type.

It's was not ideal that the Request is passed to this function, what can be considered as insecure user data, wat of course always should be validated. To fix this issue i moved the relay state property to the controller so it's given as parameter instead. 

But i don't think the controller can validate the relay state. I added some documentation why not, and what assumptions i have taken here.